spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java

```diff
@@ -733,7 +734,18 @@ else if (applicationContext != null) {
 		kafkaMessageDrivenChannelAdapter.setApplicationContext(applicationContext);
 		ErrorInfrastructure errorInfrastructure = registerErrorInfrastructure(destination,
 				consumerGroup, extendedConsumerProperties);
-		if (!extendedConsumerProperties.isBatchMode()
+
+		if (extendedConsumerProperties.getMaxAttempts() > 1 && !extendedConsumerProperties.getExtension().isEnableDlq()) {
+			if (transMan != null) {
+				messageListenerContainer.setAfterRollbackProcessor(new DefaultAfterRollbackProcessor<>(new FixedBackOff(extendedConsumerProperties.getBackOffInitialInterval(),
+						extendedConsumerProperties.getMaxAttempts() - 1)));
+			}
+			else {
+				messageListenerContainer.setCommonErrorHandler(new DefaultErrorHandler(new FixedBackOff(extendedConsumerProperties.getBackOffInitialInterval(),
+						extendedConsumerProperties.getMaxAttempts() - 1)));
+			}
+		}
+		else if (!extendedConsumerProperties.isBatchMode()
 				&& extendedConsumerProperties.getMaxAttempts() > 1
```

Comment:
No; this will still create a compound retry (3x3 by default).

We need `new FixedBackoff(0L, 0L)` until we remove the `RetryTemplate` altogether.

Reply:
I didn't look closely enough. This new `if()` block is the problem.

It prevents the `RetryTemplate` from being setup (in the next block).

I see that I already dealt with transactions in the `else` - disabling RetryTemplate and constructs an `ExponentialBackOff` from the binding properties.
(We can't use a `RetryTemplate` with transactions because it would run in the transaction, we need the retries outside of the transaction.

I believe you should remove the new if block and set the STCEH with a `FixedBackOff(0L, 0L)` in the next block (after setting the `RetryTemplate`.